scheds/rust/scx_rustlite/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,404 @@
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_QUEUE);
+	__uint(key, 0);
+	__type(value, struct dispatched_task_ctx);
+	__uint(max_entries, MAX_ENQUEUED_TASKS);
+} dispatched SEC(".maps");
+
+/*
+ * Heartbeat timer used to periodically trigger the check to run the user-space
+ * scheduler.
+ *
+ * Without this timer we may starve the scheduler if the system is completely
+ * idle and hit the watchdog that would auto-kill this scheduler.
+ */
+struct usersched_timer {
+	struct bpf_timer timer;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, struct usersched_timer);
+} usersched_timer SEC(".maps");
+
+/*
+ * Return true if the target task @p is a kernel thread.
+ */
+static inline bool is_kthread(const struct task_struct *p)
+{
+	return !!(p->flags & PF_KTHREAD);
+}
+
+/*
+ * Return true if the target task @p is the user-space scheduler.
+ */
+static inline bool is_usersched_task(const struct task_struct *p)
+{
+	return p->pid == usersched_pid;
+}
+
+/*
+ * Dispatch a task on the local per-CPU FIFO.
+ */
+static void dispatch_task_local(struct task_struct *p, u64 enq_flags)
+{
+	dbg_msg("dispatch (local): pid=%d", p->pid);
+	scx_bpf_dispatch(p, SCX_DSQ_LOCAL, slice_ns, enq_flags);
+	__sync_fetch_and_add(&nr_kernel_dispatches, 1);
+}
+
+/*
+ * Dispatch a task on the global FIFO.
+ */
+static void
+dispatch_task_global(struct task_struct *p, u64 payload, u64 enq_flags)
+{
+	/*
+	 * Tasks without a payload are dispatched directly by the kernel, tasks
+	 * with payload are dispatched by the user-space scheduler.
+	 */
+	if (!payload)
+		dbg_msg("dispatch (global): pid=%d", p->pid);
+	else
+		dbg_msg("dispatch (user): pid=%d payload=%llu", p->pid, payload);
+	scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns, enq_flags);
+	__sync_fetch_and_add(&nr_user_dispatches, 1);
+}
+
+/*
+ * Select the target CPU for a task that is ready to run.
+ */
+s32 BPF_STRUCT_OPS(rustlite_select_cpu, struct task_struct *p, s32 prev_cpu,
+		   u64 wake_flags)
+{
+	/*
+	 * Let's keep it simple for now and always return the previously used
+	 * cpu, .enqueue() will be called on this cpu and if it's idle we can
+	 * immediately dispatch the task bypassing the user-space scheduler.
+	 */
+	return prev_cpu;
+}
+
+/*
+ * Return true if the selected CPU for the task is immediately avaliable
+ * (user-space scheduler not required), false otherwise (user-space scheduler
+ * required).
+ */
+static bool is_task_cpu_available(struct task_struct *p)
+{
+	/*
+	 * Always dispatch per-CPU kthread on the same CPU, bypassing the
+	 * user-space scheduler (in this way we can to prioritize critical
+	 * kernel threads that may potentially slow down the entire system if
+	 * they are blocked for too long).
+	 */
+        if (is_kthread(p) && p->nr_cpus_allowed == 1)
+		return true;
+	/*
+	 * Keep using the same CPU if the task can only run there or if the CPU
+	 * is idle.
+	 */
+	if (p->nr_cpus_allowed == 1 ||
```

Comment:
If we're going to stay in the kernel for any percpu thread, I think we can bring it to the top and remove the && on 189?